Tabs component

sbastn opened the pull request:
This PR adds support for Vuejs components and introduces the first component: `<tabs>`. @DXTimer is planning to use this component in his use cases/tutorial series to display code in different languages. 

`<tabs>` supports any content and not only code.

## 👀 Preview

![tabs](https://user-images.githubusercontent.com/80610/113592802-eb4c2080-9635-11eb-9c68-4a41e347a40a.gif)


## 🔨 Use

### 1. Write your content

Here is the basic template to get you started:

```html
<div class="component">
  <tabs>
    <template v-slot:ruby>Hello Ruby</template>
    <template v-slot:python>Hello Python</template>
  </tabs>
</div>
```

`tabs` supports markdown by default inside each `template` slot. If you're going to write an article with lots of code samples, I suggest that you declare the `.component` class at the top of your article, use `markdown="1"` for the content that is not inside the tabs. Like the following:

```html
<div class="component" markdown="1">
# First tutorial

Here's our hello world tutorial

  <tabs>
    <template v-slot:ruby>
~~~ruby
require 'dnsimple'
~~~
    </template>

    <template v-slot:python>
**Some Python code will be written!**
_(Eventually)_
    </template>
  </tabs>

# More explanations

  <tabs>
    <template v-slot:go>
*Go content*
    </template>
    <template v-slot:foo>
_foo content_
    </template>
  </tabs>

</div>

<script src="/dist/main.js"></script>
```

### 2. Render the component

Include `<script src="/dist/main.js"></script>` at the end of the page where `<tabs>` are used. Since this is the first draft of the component, I didn't add this script to the main layout. If the usage because more than a couple of pages, we should consider moving to the layout.

dallasread commented on src/index.js:
```diff
@@ -1 +1,13 @@
+import Vue from "vue/dist/vue.esm.js"
+import Tabs from "./components/tabs.vue"
+
+// Troubleshoot with this
```
Do we want to commit this?

sbastn replied:
I'll remove it. Thanks!

dallasread reviewed (approved):
👍  Looks good, though I could not get the markdown to work (using the old Vue component, or my additions). I suspect this is a local configuration issue.

sbastn commented on Rakefile:
```diff
@@ -52,7 +66,7 @@ end
 task run: [:compile] do
   Bundler.with_unbundled_env do
-    sh("bundle exec nanoc live")
+    sh("node node_modules/concurrently/bin/concurrently.js 'bundle exec nanoc live' './node_modules/.bin/webpack --watch'")
```
Yes, will do

sbastn commented on Rakefile:
```diff
@@ -38,6 +38,20 @@ task :compile_openapi do
   File.write("output/v2/openapi.json", JSON.dump(YAML.load(data)))
 end
 
+desc "Compile Javascript assets"
```
It is

sbastn commented on content/assets/css/_tabs.scss:
```diff
@@ -0,0 +1,34 @@
+$border-color: #ccc;
```
You're right. I can do better

sbastn commented:
After thinking about the feedback, I'm going to make this component nicer. I'm going to adapt @dallasread syntax and include `markdown` capability inside the component via https://marked.js.org/ so there is no need to use `markdown="1"`

sbastn commented:
@DXTimer @dallasread I reworked the `tabs` component to do the following:
1. Support markdown by default. You no longer need to explicitly declare markdown blocks inside each tab.
2. Removed the `tab0`, `tab1` structured naming to support better names.

Looking forward to your feedback!

dallasread commented on src/components/tabs.vue:
```diff
@@ -0,0 +1,39 @@
+      }
+    },
+    mounted () {
+      this.tabs = Object.entries(this.$slots).map(slot => {
+        return slot[0]; // returns the slot name, for instance, "ruby" for v-slot:ruby
+      });
+      const options = {
+        highlight: (code) => hljs.highlightAuto(code).value,
+      };
+      this.contents = Object.entries(this.$slots).map(slot => {
+        return marked(slot[1][0].text, options); // returns the content for the slot converted to markdown
+      });
```
Would it be worth using `Object.keys` and `Object.values` (untested code)?

```suggestion
      this.tabs = Object.keys(this.$slots);
      const options = {
        highlight: (code) => hljs.highlightAuto(code).value,
      };
      this.contents = Object.values(this.$slots).map(text => {
        return marked(text, options); // returns the content for the slot converted to markdown
      });
```

sbastn replied:
👏 thanks

sbastn replied:
👏 thanks!

dallasread reviewed (approved):
👍  LGTM!

One caveat with this approach is the single-word tab title limit, but I have no ideas on how to solve that at this time.

DXTimer reviewed (approved):
Great work! :shipit:
